terraform/node_resource_destroy.go

```diff
@@ -237,19 +237,14 @@ func (n *NodeDestroyResourceInstance) EvalTree() EvalNode {
 				// Make sure we handle data sources properly.
 				&EvalIf{
 					If: func(ctx EvalContext) (bool, error) {
-						return addr.Resource.Resource.Mode == addrs.DataResourceMode, nil
+						if addr.Resource.Resource.Mode == addrs.DataResourceMode {
+							// deleting a data source is simply removing the state
+							state = nil
```

Comment:
:confused: having the `If` arm of this `EvalIf` contain a side-effect makes me a little nervous, because I've seen logic before that tries to pre-evaluate and "optimize" `EvalSequence` by resolving some of the higher-order `EvalNode` implementations early. Are you sure this is safe?

(Out of caution I'd kinda rather just have a silly `evalSetStateToNil` to put in the `Else` just to avoid introducing a new pattern into the mix here; I'm not _super_ set on it because we're hoping to phase out the `EvalNode` concept in future anyway, but I think keeping things consistent will make that later work easier to think about and plan.)

Reply:
Yeah, I can change this back. It should be safe, but there's no reason to not just use `EvalWriteState` for consistency.